module/zfs/zio.c

```diff
@@ -3661,6 +3661,17 @@ zio_dva_throttle_done(zio_t *zio)
 	ASSERT3U(zio->io_child_type, ==, ZIO_CHILD_VDEV);
 	ASSERT(vd != NULL);
 	ASSERT3P(vd, ==, vd->vdev_top);
+	/*
+	 * If fault injection is enabled, the ZIO_FLAG_IO_RETRY could be set by
+	 * zio_handle_device_injection() to generate the FMA events and update
+	 * stats. Hence, ignore the flag and process such zios.
+	 *
+	 * A better fix would be to add another flag in the zio_t to indicate
+	 * that the zio is failed because of a zinject rule. Considering the
+	 * fact that we do this in debug bits, we could do with the crude check
+	 * of zio_injection_enabled.
+	 */
```

Comment:
Normally I'm all for detailed comments but this one really breaks up the flow of this function.  Since the updated ASSERTs are largely self explanatory I think it'd be OK to simply adding this information to the commit message and remove it from here.